src/backend/distributed/master/master_cache_table_shards.c

```diff
@@ -0,0 +1,191 @@
+	{
+		ShardInterval *shardInterval = shardIntervalArray[shardIndex];
+		placementListArray[shardIndex] =
+			FinalizedShardPlacementList(shardInterval->shardId);
+	}
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		List *shardDropList = FindAbsentShardPlacementsOnWorker(workerNode,
+																shardIntervalArray,
+																placementListArray,
+																shardCount);
+		DropShardsFromWorker(workerNode, relationId, shardDropList);
+	}
+
+	pfree(placementListArray);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * FindAbsentShardPlacementsOnWorker compiles shard interval list of shards
+ * that do not have registered placement at given worker node.
+ */
+List *
+FindAbsentShardPlacementsOnWorker(WorkerNode *workerNode,
+								  ShardInterval **shardIntervalArray,
+								  List **placementListArray, int shardCount)
+{
+	List *absentShardIntervalList = NIL;
+
+	int shardIndex = 0;
+	for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+	{
+		ShardInterval *shardInterval = shardIntervalArray[shardIndex];
+		List *placementList = placementListArray[shardIndex];
+
+		ListCell *placementCell = NULL;
+		foreach(placementCell, placementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+				placementCell);
+
+			/*
+			 * Append shard interval to absent list if none of its placements is on
+			 * the worker.
+			 */
+			if (shardPlacement->nodePort == workerNode->workerPort &&
+				strncmp(shardPlacement->nodeName, workerNode->workerName,
+						WORKER_LENGTH) == 0)
+			{
+				break;
+			}
+			else if (lnext(placementCell) == NULL)
+			{
+				absentShardIntervalList = lappend(absentShardIntervalList, shardInterval);
+			}
+		}
+	}
+
+	return absentShardIntervalList;
+}
+
+
+/*
+ * DropShardsFromWorker drops provided shards belonging to a relation from
+ * given worker. It does not change any metadata at the master.
+ */
+static void
+DropShardsFromWorker(WorkerNode *workerNode, Oid relationId, List *shardIntervalList)
+{
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *relationName = get_rel_name(relationId);
+	char relationKind = get_rel_relkind(relationId);
+	StringInfo workerCommand = makeStringInfo();
+	ListCell *shardIntervalCell = NULL;
+
+	if (shardIntervalList == NIL)
+	{
+		return;
+	}
+
+	if (relationKind == RELKIND_RELATION)
+	{
+		appendStringInfo(workerCommand, DROP_REGULAR_TABLE_COMMAND, "");
+	}
+	else if (relationKind == RELKIND_FOREIGN_TABLE)
+	{
+		appendStringInfo(workerCommand, DROP_FOREIGN_TABLE_COMMAND, "");
+	}
+	else
+	{
+		ereport(ERROR, (errcode(ERRCODE_WRONG_OBJECT_TYPE),
+						errmsg("expire target is not a regular or foreign table")));
+	}
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		StringInfo shardName = makeStringInfo();
```

Comment:
I don't really like calling `makeStringInfo` in a loop here (it could result in a lot of allocations, no?)

Can you move the declaration (and `makeStringInfo()` call) outside the loop, then just call `resetStringInfo` at the end of the loop? See other places we use `resetStringInfo` within Citus for examples.